### ISSUE TITLE
Products "falling off"  the site  - after order placement

### ISSUE THREAD

**Issue by RobC1234** [Smile-SA/elasticsuite#610]
### Preconditions

Magento 2.2  EE

Magento Version : 2.2.0

ElasticSuite Version : 2.3

Environment : Production

Third party modules :

### Steps to reproduce

1. Run Full ReIndex 
2. All Products In Category displayed properly on Front-End
3. Add Item to Cart, and Place order
4.  Item Ordered no longer displays on Front-End
5. Run Full ReIndex,  Item reappears on site
(best selling items, by definition, continue to drop off the site)

### Expected result

1. Items ordered should not cause items to drop off the site

### Actual result

1. [Screenshot, logs]

**Comment by romainruaud** [comment]
Hello,

at first sight, it seems something "wrong" happens when reindexing partially products : they get reindexed after a successful order because their stock status might have changed.

Could you please list the third party modules you are using ? Maybe you have a module which is used to manage stock in a custom manner ?

Also, could you check if the same symptom happens when saving a product manually in the Back-Office ?

Regards

**Comment by romainruaud** [comment]
@RobC1234 I was not able to send you an email to the address you provided to me (Undelivered Mail Returned to Sender : Recipient address rejected) .

Can you send me an email at romain.ruaud@smile.fr to ensure that I got the proper addres ?

**Comment by romainruaud** [comment]
Ok, I think I have pointed out what is causing this issue. Imho this is linked to composite products and to a change in Magento version 2.2.

Let's describe the test case first : 

We have a grouped product (entity_id **7178**) with several childrens (73,74,78, and so on).

Indexers are on **Update on Schedule** mode.

I navigate through the website and process an order for some childrens of the grouped product (let's say 73, 77, 78 and 84).

The table _catalogsearch_fulltext_cl_ is now containing this : 

```
mysql> select * from catalogsearch_fulltext_cl;
+------------+-----------+
| version_id | entity_id |
+------------+-----------+
|          1 |        73 |
|          2 |        77 |
|          3 |        78 |
|          4 |        84 |
+------------+-----------+
```

So my 4 simple products are scheduled to be reindexed. Notice **the parent is NOT scheduled for reindexing** : probably due to the fact it's a grouped product, and I'm buying some of this product's childrens.

Then, the cronjob is coming to process reindexing of these products.

Now we go through the following piece of code : https://github.com/magento/magento2/blob/2.2/app/code/Magento/CatalogSearch/Model/Indexer/Fulltext.php#L119

Notice how it's done : 

```
            $productIds = array_unique(array_merge($ids, $this->fulltextResource->getRelationsByChild($ids)));
            $saveHandler->deleteIndex([$dimension], new \ArrayObject($productIds));
            $saveHandler->saveIndex([$dimension], $this->fullAction->rebuildStoreIndex($storeId, $ids));
```

1. Delete the products **and their parents** from the index !
2. Reindex only the childrens : they pass only `$ids` to the rebuildStoreIndex method, where they should pass `$productIds` which contains childrens AND parents.

=> They remove parents from the index, and do NOT reindex them, only the childrens...

If you look the previous 2.1 version, the code was like this : https://github.com/magento/magento2/blob/2.1/app/code/Magento/CatalogSearch/Model/Indexer/Fulltext.php#L82

```
            $dimension = $this->dimensionFactory->create(['name' => 'scope', 'value' => $storeId]);
            $saveHandler->deleteIndex([$dimension], new \ArrayObject($ids));
            $saveHandler->saveIndex([$dimension], $this->fullAction->rebuildStoreIndex($storeId, $ids));
```

I'm not sure about how this could have work properly since they were never looking for the parent products to reindex. Maybe this was broken, as suppose the following commit which is describing the change : "parent_id of deleted configurable product variation does not get sent to ElasticSearch" https://github.com/magento/magento2/commit/0b918af218117c8ab4cf2b329628486c0c5bf540 

But this is only partially fixed by Magento since they do not reindex the parents properly ....

**Comment by romainruaud** [comment]
Fixed in release 2.4.1.

Please update to this branch to fix the issue.

Do not hesitate to contact us if you are still facing issues with our module.

Thank you for your support and your enthousiasm using Elasticsuite.

Regards
